### ISSUE TITLE
Question: Is low_emission_zone handled in Valhalla?

### ISSUE THREAD

**Issue by xlqian** [valhalla/valhalla#3970]
Hello,

In order to be compliant with French/European legislation, a motorist should be notified and be aware of the fact he is traversing an LEZ(https://wiki.openstreetmap.org/wiki/Tag:boundary%3Dlow_emission_zone). 

Could you let us know if this issue has already been tackled in Valhalla? A quick search has shown nothing interesting about the LEZ. If nothing has been done regarding it yet, Are you interested in this subject? we can contribute. 

Regards

**Comment by nilsnolde** [comment]
Oh yes, I'd be. I was trying to push that years ago for openrouteservice in the German OSM community, but that discussion was so destructive that it upset me enough to just drop it.. It's still in the ether though (unfortunately only on German): https://forum.openstreetmap.org/viewtopic.php?id=65638. I think it's a quite wide topic to cover the range of low-emission-zones that exists (if that's the goal), but it surely also depends on how granular one wants to get (e.g. time-/fuel-/weight-dependencies).

On the processing side, I don't think we really do any relation processing like this, i.e. ways in polygons, but I might be wrong. It's not a show-stopper, but likely complicates things. 

If you're up for an implementation, I personally would very much welcome it and would be happy to assist here and there. Could've been a nice GSoC project as well in fact. I think it's big enough that we should discuss the process/design a bit more in detail before tackling. Would you be willing to draft an abstract of the tags and steps needed? Doesn't have to be crazy detailed, we can drill down from the bigger ones.

**Comment by kevinkreiser** [comment]
it reminds me very much of administrative boundary processing except in this case we don't store admin info we store a bit that says the edge intersected the boundary. I would suggest we follow that as a model, loosely.

the question to me is do we need to parse the boundary info at the same time we build the tiles or can we preprocess it in a separate binary like admin boundaries and add it via sqlite intersections during tile building.

maybe we could generalize valhalla_build_admins into valhalla_build_boundaries and update the sqlite schema to hold any number of interesting area features. then during admin setting in the tile build we could also set other properties based on what other area features were in there

**Comment by nilsnolde** [comment]
Hahah right, the admin boundaries.. Yeah, very similar concept. 

Then I guess we'd mostly need a more thorough review of how things are tagged in OSM and how we'd like to support those things in the graph/request. I think there's more needed than one bit, at least to get the full information, e.g. one bit to indicate there's a transition in/out of such an area, then in EdgeInfo or wherever a bit more info more about what that zone is exactly? Again, depends on what's out there and how we want to expose it in the request parameters.

**Comment by xlqian** [comment]
Thank you guys for the quick response. We appreciate it.

>  https://forum.openstreetmap.org/viewtopic.php?id=65638

I wish I could read German. :smile: 

>  we preprocess it in a separate binary like admin boundaries and add it via sqlite intersections during tile building

We have had a very similar idea which seems to be quite intuitive to us, but could be a quite time-consuming process, right? since we have to iterate every edge to figure out with which LEZ it intersects, but again, it seems quite intuitive and it's worth trying (I've no idea if what I'm saying makes any sense to you...)
 
> maybe we could generalize valhalla_build_admins into valhalla_build_boundaries ...

Sound to me like a great starting point, we're going to have a closer look at this part...

> we'd mostly need a more thorough review of how things are tagged in OSM... depends on what's out there and how we want to expose it in the request parameters.

totally agree with that, and here is what we have found regarding the tag LEZ: https://overpass-turbo.eu/s/1rkN

**Comment by nilsnolde** [comment]
> since we have to iterate every edge to figure out with which LEZ it intersects

Right, it doesn't seem intuitive, but IIRC that's what the admin processing does as well. And I don't think you really feel it too much. Also, here it's "just" crossing, not a full intersecting, which _should be_ much faster.

> totally agree with that, and here is what we have found regarding the tag LEZ: https://overpass-turbo.eu/s/1rkN

That's an easy one:) I raise you this: openstreetmap.org/relation/271888. Look at the `motor_vehicle:conditional` tag, there's quite a few variations of that. I'm not saying we want to support all of that, but unfortunately LEZ != LEZ. Where do we draw the line and how do we communicate that to the user in terms of request parameter?

**Comment by nilsnolde** [comment]
Likely also needs processing for ways: https://www.openstreetmap.org/way/113941240.

**Comment by kevinkreiser** [comment]
to really tell the full story in my mind, im kind of thinking a bigger picture architecture. what id like to see is we have a binary that pulls out area (maybe even point) features from osm that we dont strictly use for routing on:

- country/state borders for admins
- parks, beaches, geological features
- buildings, pois

and those go into spatialite with a spatial index, like we do with borders today but with different types of metadata (maybe we just save all the tag/value combos without further processing. once we have that db, we can then in graph enhancer enhance the graph, optionally, with these different things. we can add admin info to the nodes as we do today, or we can add landmark features to the edges for landmark routing, or we can do a pass to add open area features (ie invent some edges in the graph for routing over fields and parks).

im thinking about a general architecture that lets use basically pull out stuff from osm that isnt routable but can be used to enhance the routing when we build the tiles. it would then be some config option in mjolnir as to which types of features to pull out of osm in this new version of `valhalla_build_admins`.

the devil is in the details of course but i just think if we setup a general infrastructure around this we could enable a bunch of future-looking features. hell, because the format is just sqlite you could even load non-osm data in there (ignoring the licensing question at the moment) for other open datasets that arent in osm (green or scenic zones or who knows what)

**Comment by DiviyamPathak** [comment]
this would be also great in India. Unfortunately we don't have enough data and uniform legislation it would be great for MSMEs in Urban cities in India

**Comment by nilsnolde** [comment]
Anecdotally, Google Maps seems to have recently (?) implemented LEZ, but of course they have no way of respecting them in the routing:)

**Comment by nilsnolde** [comment]
Recently I thought about this again, leaving my thoughts here.. 

An alternative to the sqlite approach:

We could make them a separate structure of the tile data, instead of putting the info into directededge/edgeinfo/ The advantages would be to be much more flexible with bits & bytes

**Comment by lakovacs2** [comment]
Hi All,
 
We use Valhalla for pathfinding in our Company for a particular project.
We need to use Low Emission Zone feature, and we would like to contribute it back.

We need to realize 2 use cases for now.
1) After the path was found, we use trace_attributes of Map Matching API with the path geometry to get different attributes for the edges. 
Now we want to know also the number of the OSM id of the Low Emission Zone for every edge too.

2) We want to realize a route request feature on the Turn-by-Turn Route API to exclude all LEZs from the possible paths.
We also had some thoughts on realizing some fine grained exclusion of LEZs, but it seems to be almost impossible to implement because of the variety of LEZ rules in the different countries.

We read through the following discussion thread regarding LEZ implementation in the Valhalla project.
We really would like to align our implementation with your thoughts on this.
This is to start a discussion.
Please share your thoughts!

**Comment by schattila** [comment]
Hi @kevinkreiser , @nilsnolde ,

Please comment on the previous entry!

Thanks and regards,
Attila

**Comment by nilsnolde** [comment]
I'd be happy if someone took a stab at this!

I guess my first question here would be: where do you want to store the data in the tiles and how do you plan to access it during expansion/costing? 

> We also had some thoughts on realizing some fine grained exclusion of LEZs, but it seems to be almost impossible to implement because of the variety of LEZ rules in the different countries.

Hehe, yeah for sure.

A bit more description for the technical plan would be appreciated. I'll try to raise this on our next weekly. There's currently very little time as GSoC etc is going on, but we'll get there..

**Comment by schattila** [comment]
Hi Nils,

The plan is the following:
We plan to create a zones.sqlite file with the following tables:
zones with the following columns:
name                         TEXT         original name
name_en                   TEXT         english name (if available)
parent_zone             INTEGER  index of the parent zone (if available)
restriction_indexes BLOB        list of 32 bit indexes of the zone_access table
description               TEXT
geom                        BLOB        MULTIPOLYGON
zone_access with the following columns:
value                         BLOB       uint64_t (if available)
type                           INTEGER baldr::AccessType
modes                      INTEGER 12 bit access modes bitmask (e.g. kAutoAccess)
We need the zone_access table, because we can have nested Low Emission Zones, and it also can have multiple restriction too (like Mo-Th 07:00-20:00; Fr 07:00-15:00).

Then from zoned.sqlite, during the valhalla map creation, we plan to create an AccessRestriction based on the LEZ. In this restriction we would store the OSM id of the Low Emission Zone. This way we can get the OSM id when using the Trace Attributes API, and also could use it to restrict all LEZs, when instructed in the costing option.

**Comment by nilsnolde** [comment]
ok, cool, thanks. I'd have a few comments:
- we're about to introduce area handling properly in https://github.com/valhalla/valhalla/pull/6127 (GSoC 2026 project), so you might re-think your sqlite idea
- how do you plan to tag the edges in those LEZ areas? if every single one is tagged, it makes urban tiles (and sadly **all** hierarchies) really big! e.g. stuttgart or so. and many many access restrictions will cause urban queries to be extra slow.

EDIT: tile size is not the worry here. query performance for the feature is. oh sorry, even worse, this would affect _all_ queries, bcs the binary search space increases. we have to be very careful what we put in access restrictions (as with all other structures too..).

**Comment by nilsnolde** [comment]
hahah ok I already sketched it out.. never mind the "how do you plan" question;)

**Comment by schattila** [comment]
1) About https://github.com/valhalla/valhalla/pull/6127, is this code complete for the area handling? We have checked, but it seems, that the created structures are not used yet.
2) Do you have another mechanism, how to restrict access for edges in LEZs? As you know we have 2 use cases. First give back OSM LEZ ids for every edge in LEZ. Second is to restrict path finding in all LEZs.

**Comment by nilsnolde** [comment]
> About https://github.com/valhalla/valhalla/pull/6127, is this code complete for the area handling? We have checked, but it seems, that the created structures are not used yet.

right, it's not yet. however, I'd like to wait with a decision on how we do it for that PR before we decide here. might be that sqlite is a fair approach, we're not sure yet ourselves over there. we try to design it in a way that other polygon use cases (there's many more involving polygons!) are easily added.

> Do you have another mechanism, how to restrict access for edges in LEZs? As you know we have 2 use cases. First give back OSM LEZ ids for every edge in LEZ. Second is to restrict path finding in all LEZs.

ah, didn't _really_ have that mind still. restrict path finding is fine, you can do it as I earlier suggested. for your LEZ ID on every edge: na, that can be done in a different way. and I mean entirely different, not even involving valhalla. I mean, in theory, there could be a case for smth like a "polygon" structure in the tiles (just came to me, didn't think hard on it), but that's out-of-scope here. I generally didn't think too much on how to best extract LEZ OSM IDs from a _final_ path (so def no need to live on directed edges!). I'm mostly saying, directededge.cc or access restrictions can't be the place where any of this lives. see e.g. [my previous comment](https://github.com/valhalla/valhalla/issues/3970#issuecomment-1793572773) from 3 years ago (not how I'd do it today maybe, but not sure..).
